File: dclib/core/src/main/java/dclib/epf/graphics/EntitySpriteDrawer.java
package dclib.epf.graphics;

import java.util.ArrayList;
import java.util.Collections;
import java.util.Comparator;
import java.util.List;

import com.badlogic.gdx.graphics.g2d.PolygonSpriteBatch;

import dclib.epf.Entity;
import dclib.epf.EntityAddedListener;
import dclib.epf.EntityManager;
import dclib.epf.parts.SpritePart;
import dclib.epf.parts.TransformPart;
import dclib.graphics.ScreenHelper;

public final class EntitySpriteDrawer implements EntityDrawer {

	private final PolygonSpriteBatch spriteBatch;
	private final ScreenHelper screenHelper;
	// Don't draw new entities because their transforms might not have been initialized
	private final List<Entity> newEntities = new ArrayList<Entity>();

	public EntitySpriteDrawer(final PolygonSpriteBatch spriteBatch, final ScreenHelper screenHelper,
			final EntityManager entityManager) {
		this.spriteBatch = spriteBatch;
		this.screenHelper = screenHelper;
		entityManager.listen(new EntityAddedListener() {
			@Override
			public void added(final Entity entity) {
				newEntities.add(entity);
			}
		});
	}

	@Override
	public final void draw(final List<Entity> entities) {
		final EntityZComparator entityZComparator = new EntityZComparator();
		Collections.sort(entities, entityZComparator);
		screenHelper.setProjectionMatrix(spriteBatch);
		spriteBatch.begin();
		for (Entity entity : entities) {
			if (newEntities.contains(entity)) {
				newEntities.remove(entity);
			} else {
				draw(entity);
			}
		}
		spriteBatch.end();
	}

	private void draw(final Entity entity) {
		SpritePart spritePart = entity.tryGet(SpritePart.class);
		if (spritePart != null) {
			spritePart.getSprite().draw(spriteBatch);
		}
	}

	private static class EntityZComparator implements Comparator<Entity> {

		@Override
		public final int compare(final Entity e1, final Entity e2) {
			return Float.compare(getValue(e1), getValue(e2));
		}

		private float getValue(final Entity entity) {
			TransformPart transformPart = entity.tryGet(TransformPart.class);
			if (transformPart != null) {
				return transformPart.getTransform().getZ();
			} else {
				return 0;
			}
		}

	}

}

File: dclib/core/src/main/java/dclib/epf/graphics/EntitySpriteDrawer.kt
package dclib.epf.graphics

import com.badlogic.gdx.graphics.g2d.PolygonSpriteBatch
import dclib.epf.Entity
import dclib.epf.EntityManager
import dclib.epf.parts.SpritePart
import dclib.epf.parts.TransformPart
import dclib.graphics.ScreenHelper
import dclib.map.MapController
import dclib.map.MapUtils
import java.util.ArrayList

class EntitySpriteDrawer(
		private val spriteBatch: PolygonSpriteBatch,
		private val screenHelper: ScreenHelper,
		private val mapController: MapController,
		private val getEntities: () -> List<Entity>,
		entityManager: EntityManager
) : EntityDrawer {
	private val FOREGROUND_Z = 0f

	// Ensures new entities aren't drawn because their transforms might not have been initialized
	private val newEntities = ArrayList<Entity>()

	init {
		entityManager.entityAdded.on { newEntities.add(it.entity) }
	}

	override fun getName(): String {
		return "sprite"
	}

	override fun draw(entities: Collection<Entity>) {
		val drawnEntities = getEntities()
//		mapController.renderLayer(MapUtils.BACKGROUND_INDEX)
//        draw(drawnEntities, Float.NEGATIVE_INFINITY, FOREGROUND_Z)
		mapController.renderLayer(MapUtils.FOREGROUND_INDEX)
		draw(drawnEntities, FOREGROUND_Z, Float.POSITIVE_INFINITY)
	}

    private fun draw(entities: Collection<Entity>, minZ: Float, maxZ: Float) {
		val filteredEntities = entities.filter {
			val z = it[TransformPart::class].transform.z
			z >= minZ && z < maxZ
		}
        screenHelper.setProjectionMatrix(spriteBatch)
        spriteBatch.begin()
        for (entity in filteredEntities) {
            if (newEntities.contains(entity)) {
                newEntities.remove(entity)
            } else {
                draw(entity)
            }
        }
        spriteBatch.end()
    }

	private fun draw(entity: Entity) {
		val spritePart = entity.tryGet(SpritePart::class)
		if (spritePart != null) {
			spritePart.sprite.draw(spriteBatch)
		}
	}
}